libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -121,22 +123,51 @@ string IRGenerationContext::newYulVariable()
 	return "_" + to_string(++m_varCounter);
 }
 
-string IRGenerationContext::generateInternalDispatchFunction(Arity const& _arity)
+void IRGenerationContext::initializeInternalDispatch(InternalDispatchMap _internalDispatch)
 {
-	// UNIMPLEMENTED: Internal library calls via pointers are not implemented yet.
-	// We're not generating code for internal library functions here even though it's possible
-	// to call them via pointers. Right now such calls end up triggering the `default` case in
-	// the switch above.
-	set<FunctionDefinition const*> functions;
-	for (auto const& contract: mostDerivedContract().annotation().linearizedBaseContracts)
-		for (FunctionDefinition const* function: contract->definedFunctions())
-			if (!function->isConstructor() && Arity::fromDefinition(*function) == _arity)
-			{
-				functions.insert(function);
-				enqueueFunctionForCodeGeneration(*function);
-			}
+	solAssert(internalDispatchClean(), "");
+
+	for (set<FunctionDefinition const*> const& functions: _internalDispatch | boost::adaptors::map_values)
+		for (auto function: functions)
+			enqueueFunctionForCodeGeneration(*function);
+
+	m_internalDispatchMap = move(_internalDispatch);
+}
 
-	return generateInternalDispatchFunction(_arity, move(functions));
+InternalDispatchMap IRGenerationContext::consumeInternalDispatchMap()
+{
+	m_directInternalFunctionCalls.clear();
+
+	InternalDispatchMap internalDispatch = move(m_internalDispatchMap);
+	m_internalDispatchMap.clear();
+	return internalDispatch;
+}
+
+void IRGenerationContext::internalFunctionCalledDirectly(Expression const& _expression)
+{
+	solAssert(m_directInternalFunctionCalls.count(&_expression) == 0, "");
+
+	m_directInternalFunctionCalls.insert(&_expression);
+}
+
+void IRGenerationContext::internalFunctionCalled(Expression const& _expression)
+{
+	solAssert(IRHelpers::referencedFunctionDeclaration(_expression), "");
+
+	if (m_directInternalFunctionCalls.count(&_expression) == 0)
+	{
+		FunctionDefinition const& resolvedFunction = IRHelpers::referencedFunctionDeclaration(_expression)->resolveVirtual(mostDerivedContract());
```

Comment:
Are these always virtual calls?

Reply:
Right. I'll change it.

Reply:
Good catch. They are not. But I'm not sure how it is passing tests then. I found this problem earlier and it was failing the test that calls a contract method from the base class and I was sure I fixed it. I need to take a closer look at what I did there.

Reply:
I know what happened. I had two similar problems with virtual resolution, one affecting only direct calls and the other only calls via pointers. I caught the former in tests because that's a more common case. The one pointed out here was not covered.

I fixed my code and also added a test that would detect such a situation (`inherited_function_via_pointer.sol`).